### ISSUE TITLE
Handoff from Vanilla JS App to Angular App Authentication

### ISSUE THREAD

**Issue by roopeshchinnakampalli** [AzureAD/microsoft-authentication-library-for-js#5407]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.24.0

### Wrapper Library

MSAL Angular (@azure/msal-angular)

### Wrapper Library Version

2.4.5

### Public or Confidential Client?

Public

### Description

Let's assume, We've an Angular app with MSAL Authentication. When we load this app in the browser, it has to download all the required JS files before application starts showing up and then the authentication kicks in, where it redirects to the Azure AD and come back and reload the app.

So Authentication is delayed till the source application JS is downloaded and parsed. 

I'm thinking is there any way, we could have the plain MSAL.js library injected in to the index.html of the Angular app, which allows to kick in the authentication even before the Angular app Javascript files download.

This way, we can achieve early authentication and avoid redudant download of Angular app Javascript files.

Question:

How this handoff will happen from the authentication happened from the MSAL.js to authentication from Angular (@azure/msal-angular) library.

I've tried this and I see issues like `Browser Auth Error` & `Interaction Error` errors. Looks like there's no formal handshake between the two modes of authentication even though we use the same cache location for storage of tokens.

Any thoughts are appreciated.

### Identity Provider

Azure AD / MSA

### Source

External (Customer)

**Comment by tnorling** [comment]
> I've tried this and I see issues like Browser Auth Error & Interaction Error errors. Looks like there's no formal handshake between the two modes of authentication even though we use the same cache location for storage of tokens.

Can you please go into more detail about this? What exactly did you try and what happened? 

I'm also not sure I understand this statement: "Looks like there's no formal handshake between the two modes of authentication". msal-angular is merely a thin wrapper around msal-browser to provide Angular features like Interceptors and Guards but all operations still go through msal-browser under the hood.

**Comment by roopeshchinnakampalli** [comment]
I've downloaded one of your sample application of Angular implementing the MSAL auth. So, in the index.html file, I've added the code from the sample app for Vanilla JS implementation. To be precise, I've this below code - 

`
  <script type="text/javascript" src="https://alcdn.msauth.net/browser/2.31.0/js/msal-browser.min.js"></script>

    <script>

      const msalConfig = {
    auth: {
      authority: 'https://login.microsoftonline.com/',
      clientId: 'xxxxxxxxxxxxxxxx',
      redirectUri: 'http://localhost:4200',
      postLogoutRedirectUri: 'xxxxxxxxxxxxxxxx',
      navigateToLoginRequestUrl: true
    },
    cache: {
      cacheLocation: 'sessionStorage'
    },
    logs: false
  };

  const loginRequest = {
    scopes: ["User.Read"]
};


let signInType;
let accountId = "";

// Create the main myMSALObj instance
// configuration parameters are located at authConfig.js
const myMSALObj = new msal.PublicClientApplication(msalConfig);

// Redirect: once login is successful and redirects with tokens, call Graph API
myMSALObj.handleRedirectPromise().then(handleResponse).catch(err => {
    console.error(err);
});

function handleResponse(resp) {
    if (resp !== null) {
        accountId = resp.account.homeAccountId;
        myMSALObj.setActiveAccount(resp.account);
    } else {
        // need to call getAccount here?
        const currentAccounts = myMSALObj.getAllAccounts();
        if (!currentAccounts || currentAccounts.length < 1) {
            return;
        } else if (currentAccounts.length > 1) {
            // Add choose account code here
        } else if (currentAccounts.length === 1) {
            const activeAccount = currentAccounts[0];
            myMSALObj.setActiveAccount(activeAccount);
            accountId = activeAccount.homeAccountId;
        }
    }
}
if (!myMSALObj.getAllAccounts().length) {
// Sign In
  myMSALObj.loginRedirect(loginRequest);
}


    </script>
`

When I do this i'm thrown with this errors in the console - 

**VM451641:106 ClientAuthError: state_not_found: State not found:  Cached State**

Basically what I mean by handshake is - "If we setup a plain vanilla auth in the Angular index.html and after initializing it again in Angular app once its bootstrapped, will it work seamlessly? To my knowledge, since idtoken is already dropped inside the browser it should work".

Also, for Angular App, we add - <app-redirect /> tag beside the <app-root>. Will that cause any harm?

**Comment by tnorling** [comment]
There's no technical reason this shouldn't work, however, you should be careful that `handleRedirectPromise` is not called both in the VanillaJS and Angular app at the same time, that's likely what's causing the error here. If you refresh the page after getting this error does it work?

I'm not sure how you would delay Angular bootstrap until after the VanillaJS script has completed but that would be the way to resolve this.

**Comment by roopeshchinnakampalli** [comment]
Yes. That's what I was checking on the handoff mechanism which allows the Angular app to wait till the Vanilla JS code is done.

I will do some more tests and post the result here. If we can get this working, We have a very good authentication setup for SPA (React, Angular etc.,)
